### PR TITLE
Adding feature for subscription tier update in an existing subscription

### PR DESCRIPTION
**Pull request by CrowleyRajapakse**
## Issue

resolves wso2/product-apim#5246

### Use Case
Provide capability to change subscription tier without having to delete the subscription and resubscribing to the same API.

#### REST API Changes
- Added a new field to Subscription Resource

	requestedThrottlingPolicy:
 		type: string
example: Unlimited

- Added PUT HTTP method to '/subscriptions/{subscriptionId}' endpoint

"PUT https://localhost:9443/api/am/store/v1/subscriptions/5b65808c-cdf2-43e1-a695-de63e3ad0ae9\nAuthorization: Bearer ae4eae22-3f65-387b-a171-d37eaa366fa8\n\n{\n    \"throttlingPolicy\": \"Gold\",\n  \"requestedThrottlingPolicy\": \"Unlimited\",\n    \"apiId\": \"c43a325c-260b-4302-81cb-768eafaa3aed\",\n    \"applicationId\": \"c30f3a6e-ffa4-4ae7-afce-224d1f820524\"\n    \"status\": \"UNBLOCKED\"\n}"

#### Database Changes
- Added a new column, TIER_ID_PENDING to AM_SUBSCRIPTIONS table. This is to store the requested tier without losing the details regarding the already subscribed tier.

#### Workflow Changes

- Added a new workflow extension

`<SubscriptionUpdate executor="org.wso2.carbon.apimgt.impl.workflow.SubscriptionUpdateSimpleWorkflowExecutor"/>
    `

#### Impl Changes

Added a new subscription status TIER_UPDATE_PENDING, this is to the state where previous state is UNBLOCKED and then a subscription tier update has been triggered and currently in the approval process but not yet approved. Subscription which is in this state is also in UNBLOCKED state for the current existing tier. So users can use the  same previous subscription tier just as before without having to wait for new tier approval.

#### UI Changes

- All the Subscription states.
Screenshot 2020-04-29 at 13 08 32

- Subscription which is in UNBLOCKED state can be edited.
Screenshot 2020-04-29 at 09 13 34

Screenshot 2020-04-29 at 09 14 06

- Subscription which is in TIER_UPDATE_PENDING can be edited, which means we are requesting another new tier. Here, we are showing previous tier as well as the requested tier.
Screenshot 2020-04-29 at 09 14 43

- Subscription which is in BLOCKED state can’t be edited for tier.
Screenshot 2020-04-29 at 09 14 59

- Subscription which is in ON_HOLD state can’t be edited for tier.
Screenshot 2020-04-29 at 09 15 17

- Subscription which is in REJECTED state can’t be edited for tier.
Screenshot 2020-04-29 at 13 08 51

- Workflows Extension details can be edited by using the carbon console.

Screenshot 2020-04-29 at 11 53 03

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java [hunk 1]
```diff
@@ -466,11 +468,25 @@
...
+                    "   AM_SUBSCRIPTION " +
+                    " WHERE " +
+                    "   UUID = ? ";
+
     public static final String ADD_SUBSCRIPTION_SQL =
             " INSERT INTO " +
             "   AM_SUBSCRIPTION (TIER_ID,API_ID,APPLICATION_ID,SUB_STATUS,SUBS_CREATE_STATE,CREATED_BY,CREATED_TIME, " +
-                    "UPDATED_TIME, UUID) " +
-            " VALUES (?,?,?,?,?,?,?,?,?)";
+                    "UPDATED_TIME, UUID, TIER_ID_PENDING) " +
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java:482]: formatting

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/SubscriptionUpdateSimpleWorkflowExecutor.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                        APIConstants.SubscriptionStatus.UNBLOCKED, subscriptionWorkflowDTO.getRequestedTierName());
+            } else if (subscriptionWorkflowDTO.getStatus() == WorkflowStatus.CREATED ||
+                    subscriptionWorkflowDTO.getStatus() == WorkflowStatus.REGISTERED) {
+                apiMgtDAO.updateSubscriptionStatus(Integer.parseInt(subscriptionWorkflowDTO.getWorkflowReference()),
+                        APIConstants.SubscriptionStatus.TIER_UPDATE_PENDING);
+            } else if (subscriptionWorkflowDTO.getStatus() == WorkflowStatus.REJECTED) {
+                apiMgtDAO.updateSubscriptionStatusAndTier(Integer.parseInt(subscriptionWorkflowDTO.getWorkflowReference()),
+                        APIConstants.SubscriptionStatus.UNBLOCKED, subscriptionWorkflowDTO.getTierName());
+            }
+
```
Comment (fazlan-nazeem) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/SubscriptionUpdateSimpleWorkflowExecutor.java:106]: remove newline

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/SubscriptionUpdateWSWorkflowExecutor.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+            auth.setAuthSchemes(authSchemes);
+
+            if (contentType == null) {
+                options.setProperty(Constants.Configuration.MESSAGE_TYPE, HTTPConstants.MEDIA_TYPE_TEXT_XML);
+            }
+            options.setProperty(HTTPConstants.AUTHENTICATE, auth);
+            options.setManageSession(true);
+        }
+        client.setOptions(options);
+
```
Comment (fazlan-nazeem) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/SubscriptionUpdateWSWorkflowExecutor.java:200]: remove unwanted newlines

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/WorkflowExecutorFactory.java [hunk 4]
```diff
@@ -113,6 +113,9 @@ public WorkflowDTO createWorkflowDTO(String wfType) {
         }else if(WorkflowConstants.WF_TYPE_AM_SUBSCRIPTION_CREATION.equals(wfType)){
             workflowDTO = new SubscriptionWorkflowDTO();
             workflowDTO.setWorkflowType(wfType);
+        }else if(WorkflowConstants.WF_TYPE_AM_SUBSCRIPTION_UPDATE.equals(wfType)){
```
Comment (fazlan-nazeem) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/WorkflowExecutorFactory.java:116]: formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/SubscriptionsApiServiceImpl.java [hunk 5]
```diff
@@ -244,6 +244,107 @@ public Response subscriptionsPost(SubscriptionDTO body, String xWSO2Tenant, Mess
...
+                addedSubscriptionDTO.setRedirectionParams(payload);
+            }
+
+            return Response.ok(new URI(RestApiConstants.RESOURCE_PATH_SUBSCRIPTIONS + "/" +
+                    addedSubscribedAPI.getUUID())).entity(addedSubscriptionDTO).build();
+
+        } catch (APIMgtAuthorizationFailedException e) {
+            //this occurs when the api:application:tier mapping is not allowed. The reason for the message is taken from
+            // the message of the exception e
+            RestApiUtil.handleAuthorizationFailure(e.getMessage(), e, log);
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/SubscriptionsApiServiceImpl.java:329]: use exceptionCodes

File: features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql [hunk 6]
```diff
@@ -1341,6 +1342,7 @@ CREATE TABLE IF NOT EXISTS AM_API_PRODUCT_MAPPING (
 CREATE TABLE IF NOT EXISTS AM_SUBSCRIPTION (
     SUBSCRIPTION_ID INTEGER AUTO_INCREMENT,
     TIER_ID VARCHAR(50),
+    TIER_ID_PENDING VARCHAR(50),
```
Comment (tharindu1st) [inline features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql:1345]: Add to other database types

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/SubscriptionUpdateSimpleWorkflowExecutor.java (hunk 2 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/SubscriptionUpdateSimpleWorkflowExecutor.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/SubscriptionUpdateWSWorkflowExecutor.java (hunk 3 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/SubscriptionUpdateWSWorkflowExecutor.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/WorkflowExecutorFactory.java (hunk 4 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/WorkflowExecutorFactory.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java (hunk 1 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/SubscriptionsApiServiceImpl.java (hunk 5 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/SubscriptionsApiServiceImpl.java:329]: It seems this APIMgtAuthorizationFailedException is handled like same as this in all other usages.

File: features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql (hunk 6 above)
Comment (CrowleyRajapakse) [reply features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql:1345]: resolved

**Comment by bhathiya** [comment]
Please fix the messages. 

examples: 

image

image

**Comment by bhathiya** [comment]
> Please fix the messages.
> 
> examples:
> 
> image
> 
> image

@shaniR can you help?

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Applications/Details/SubscriptionTableData.jsx [hunk 7]
```diff
@@ -158,9 +238,103 @@ class SubscriptionTableData extends React.Component {
...
+                                                            <FormattedMessage
+                                                                defaultMessage={`Assign a new Throttling policy tier to the existing subscription`}
+                                                                id='Applications.Details.SubscriptionTableData.update.throttling.policy.helper'
+                                                            />
+                                                        )}
+                                                        margin='normal'
+                                                        variant='outlined'
+                                                    >
+                                                        {this.state.tiers.map((tier) => (
+                                                            <MenuItem key={tier.value} value={tier.value}>
```
Comment (chanaka3d) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Applications/Details/SubscriptionTableData.jsx:302]: Make sure this key value is a unique value for each. If the key is duplicated, there can be unpredictable issues.

Comment (CrowleyRajapakse) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Applications/Details/SubscriptionTableData.jsx:302]: Yes, in here, the tier key value pair is unique.

**Pull request merged**